packages/core/src/routes/interaction/utils/sign-in-experience-validation.ts

```diff
@@ -0,0 +1,131 @@
+import type { Event, SignInExperience, Profile } from '@logto/schemas';
+import { SignUpIdentifier, SignInMode, SignInIdentifier } from '@logto/schemas';
+
+import RequestError from '#src/errors/RequestError/index.js';
+import assertThat from '#src/utils/assert-that.js';
+
+import type { IdentifierPayload } from '../types/guard.js';
+
+const forbiddenEventError = new RequestError({ code: 'auth.forbidden', status: 403 });
+
+const forbiddenIdentifierError = new RequestError({
+  code: 'user.sign_in_method_not_enabled',
+  status: 422,
+});
+
+export const signInModeValidation = (event: Event, { signInMode }: SignInExperience) => {
+  if (event === 'sign-in') {
+    assertThat(signInMode !== SignInMode.Register, forbiddenEventError);
+  }
+
+  if (event === 'register') {
+    assertThat(signInMode !== SignInMode.SignIn, forbiddenEventError);
+  }
+};
+
+export const identifierValidation = (
+  identifier: IdentifierPayload,
+  signInExperience: SignInExperience
+) => {
+  const { signIn, signUp } = signInExperience;
+
+  // Username Password Identifier
+  if ('username' in identifier) {
+    assertThat(
+      signIn.methods.some(
+        ({ identifier: method, password }) => method === SignInIdentifier.Username && password
+      ),
+      forbiddenIdentifierError
+    );
+
+    return;
+  }
+
+  // Email Identifier
+  if ('email' in identifier) {
+    assertThat(
+      // eslint-disable-next-line complexity
+      signIn.methods.some(({ identifier: method, password, verificationCode }) => {
+        if (method !== SignInIdentifier.Email) {
+          return false;
+        }
```

Comment:
Maybe we can break it down and assert at smaller scope, and return more user friendly error messsages? For example, ditch the outer assertThat, and do it internally in the `.some` iteration.

```ts
signIn.methods.some((...) => {
  assertThat(method === SignInIdentifier.Email, MissingEmailError)
  assertThat('password' in identifier && !password, MissingPasswordError)
})
```

Reply:
The SIE guard is different compared to the body payload guard.  After passing through the payload guard, it will be a valid identifier data format. This middleware, it will verify if the target identifier + verification method is enabled or not. So the error will be a general one, `sign_in_method_not_enabled`